### DISCUSSION TITLE
Get a Component from inside an Entity's event callback closure?

### DISCUSSION THREAD

**Discussion by wbpaley** [bevyengine/bevy#11841] [Q&A] [answered]
Hi all--thanks for the great help so far!

This time I have several pointers hovering over entities and need to count how many are over each.

Think a usual case of multi-touch. Although right now I'm creating a small cloud of pointers around the cursor-driven standard one so I can do a "sloppy line select", i.e. can hover near the line and still arm it.

(Alternatively, maybe there's a non-skinny-ray selection mechanism similar to Rapier's shape-casting?)

Anyway, with a single pointer I could insert & delete a marker component. (The first code block below)

```
#[derive(Component)]
struct Hovered {
    entity: Entity,
    count: u32,
}            
...

commands.spawn((
      //  mesh, etc. elided
      PickableBundle::default(),
      On::<Pointer<Over>>::commands_mut(|event, commands| {
          commands.entity(event.target).insert(Hovered { entity: event.target, count: 1 });
      }),
      On::<Pointer<Out>>::commands_mut(|event, commands| {
          commands.entity(event.target).remove::<Hovered>()
      }),
  ));
```

I know I can get access to the component fields to do this increment/decrement in a normal system, but would prefer to localize this trivial increment/decrement in the entity callback itself. Made-up pseudocode calls in this second code-block. Is it possible?

```
commands.spawn((
      //  mesh, etc. elided
      PickableBundle::default(),
      On::<Pointer<Over>>::commands_mut(|event, commands| {
          // insert_if_absent() here is my made-up call 
          // If there's something like the get_mut() below and it returns None
          // I can add the component, but cleaner if I didn't have to--
          // like I think get_or_spawn() does for us elsewhere
          commands.entity(event.target).insert_if_absent(Hovered { entity: event.target, count: 0 });
          // get_mut() below is my main made-up call--is there something that actually does this?
          commands.entity(event.target).get_mut::<Hovered>().unwrap().count += 1;
      }),
      On::<Pointer<Out>>::commands_mut(|event, commands| {
          commands.entity(event.target).get_mut::<Hovered>().unwrap().count -= 1;
      }),
  ));

```

**Comment by SkiFire13** [comment, marked as answer]
You cannot directly access to components using `Commands`, but in this case you don't have to necessarily use `Commands`. `On` has other constructors, in particular there is `target_component_mut` if you only want to mutate a component on the target (like in your `On::<Pointer<Out>>` example), and `run` if you want to run a full system (which you might want to do in your `On::<Pointer<Over>>` example, also note that in this case you'll need to manually get the target data using a `Listener<E>` system parameter).

**Reply by wbpaley** [reply]
Thanks! Lovely. Easy.

(Perhaps the first bit of surprise-conciseness I've come across in my two-month Rust/Bevy journey...  ; )

Code snippet for even-newer-than-me newbies:
```
commands.entity(id)
    .insert((
        HoverCount {
            entity: id,
            count: 0,
        },
        On::<Pointer<Over>>::target_component_mut::<HoverCount>(|over, hover_count| {
            hover_count.count += 1;
            println!("{:?} {}", over.target, hover_count.count);
        }),
        On::<Pointer<Out>>::target_component_mut::<HoverCount>(|out, hover_count| {
            hover_count.count -= 1;
        }),
```
